ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,429 @@
+   *                   be created. This can be {@code null} if pooling is disabled.
+   * @param requiredSwapSegmentsPerSize the number of swap segments needed for each segment size in the pool.
+   * @param metrics a {@link StorageManagerMetrics} instance.
+   * @throws StoreException
+   */
+  DiskSpaceAllocator(boolean enablePooling, File reserveDir, long requiredSwapSegmentsPerSize,
+      StorageManagerMetrics metrics) {
+    this.enablePooling = enablePooling;
+    this.reserveDir = reserveDir;
+    this.requiredSwapSegmentsPerSize = requiredSwapSegmentsPerSize;
+    this.metrics = metrics;
+    try {
+      if (enablePooling && reserveDir.isDirectory()) {
+        reserveFiles = inventoryExistingReserveFiles();
+        poolState = PoolState.INVENTORIED;
+      }
+    } catch (Exception e) {
+      logger.error("Could not inventory preexisting reserve directory", e);
+    }
+  }
+
+  /**
+   * This method will initialize the pool such that it matches the {@link DiskSpaceRequirements} passed in.
+   * The following steps are performed:
+   * <ol>
+   *   <li>
+   *     The requirements list is aggregated into an overall requirements map. This stage will ensure that the swap
+   *     segments requirements per each segment size described in a DiskSpaceRequirements struct (even if a blob store
+   *     needs 0 additional non-swap segments). The number of swap segments needed is calculated by counting the total
+   *     number of swap segments allocated to the stores (as described by the disk space requirements). If the number
+   *     of swap segments allocated is less than the required swap segment count (passed in via configuration) for a
+   *     certain segment size, the difference is added into the overall requirements map. This ensures that the total
+   *     (allocated and pooled) swap segment count equals the required count.
+   *   </li>
+   *   <li>
+   *     Unneeded segments are deleted. These are segments that were present in the pool on disk, but are not needed
+   *     according to the overall requirements.
+   *   </li>
+   *   <li>
+   *     Any additional required segments are added. For each segment size, additional files will be preallocated to
+   *     until the number of files at that size match the overall requirements map. If fallocate is supported on the
+   *     system, and the call fails, pool initialization will fail.
+   *   </li>
+   * </ol>
+   * WARNING: No calls to {@link #allocate(File, long)} and {@link #free(File, long)} may occur while this method is
+   *          being executed.
+   * @param requirementsList a collection of {@link DiskSpaceRequirements}s objects that describe the number and
+   *                         segment size needed by each store.
+   * @throws StoreException if the pool could not be allocated to meet the provided requirements
+   */
+  void initializePool(Collection<DiskSpaceRequirements> requirementsList) throws StoreException {
```

Comment:
do we have checks to verify that overAllRequirements doesn't surpass the disk capacity. Feel free to either create an issue and add it later or if its trivial, you can add it in this patch itself.

Reply:
I will create an issue to add this soon. We can discuss the options offline